sale_order_product_assortment/models/sale_order.py

```diff
@@ -37,3 +38,39 @@ def _compute_product_assortment_ids(self):
                     )
             if self.allowed_product_ids:
                 self.has_allowed_products = True
+
+    @api.model
+    def create(self, vals):
+        res = super().create(vals)
+        res._validate_allowed_products()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        self._validate_allowed_products()
+        return res
+
+    def _validate_allowed_products(self):
+        """
+        Check if all the products in the order lines
+        contains products that are allowed for the partner
+        """
```

Comment:
Is it possible to make this using @api.constrains instead of overwriting create and write?

Reply:
Good point, thank you.